src/app/accounts/update-account-status.ts

```diff
@@ -0,0 +1,17 @@
+import { AccountsRepository } from "@services/mongoose"
+
+export const updateAccountStatus = async ({
+  id,
+  status,
+}: {
+  id: string
+  status: string
+}): Promise<Account | Error> => {
+  const accountsRepo = AccountsRepository()
+
+  const account = await accountsRepo.findById(id as AccountId)
+  if (account instanceof Error) return account
+
+  account.status = status as AccountStatus
```

Comment:
account status needs a `checkedToXxx` method. It is not safe to cast a `string` as an `AccountStatus`.

Reply:
the check is already done at the [graphql level](https://github.com/GaloyMoney/galoy/blob/521310ac85202eb0158a6fca72d33b3b000418d5/src/graphql/admin/root/mutation/user-update-status.ts#L14). maybe casting should happen at the graphql layer?

otherwise doing the check here would mean we're looking at the validity of status enum twice

Reply:
that is a different type. GraphQL layer has its own AccountStatus:
```
import AccountStatus from "@graphql/admin/types/scalar/account-status"
```

Reply:
The graphql layer check gives an early return to clients (at query parsing time) - the domain layer check is authoritative.

Reply:
The GQL layer definition should reference the domain definition 
here is the source of truth definition: https://github.com/GaloyMoney/galoy/blob/main/src/domain/accounts/index.ts#L10-L13
Look how we are doing it here:
https://github.com/GaloyMoney/galoy/blob/main/src/graphql/types/scalar/tx-status.ts#L7-L9

Reply:
I guess i don't understand why we want to make a distinction between the `graphql` and `domain` layer. seems we are making twice the work here: ie checking twice for the same thing, with a lot of code duplication.

Reply:
Graphql = coupled to clients shouldn't break
Domain = coupled to business rules - will evolve -> breaking changes

Their lifecycles, reasons to change is different. If we don't decouple them we cannot evolve our domain logic without causing breaking changes to clients (which we don't want for API v2).

Of course keeping them in sync as much as possible is useful to minimize cognitive overhead - but coupling them will significantly slow us down in the mid term unless we are also willing to push breaking changes to GQL.

Reply:
> Decoupled domain layer is a great asset to have in my experience.
not my experience so far. I have lost 1d on trying to get my head [around this issue](https://github.com/GaloyMoney/galoy/issues/781). 

Per this [book](https://book.productionreadygraphql.com/), Facebook (creator of GraphQL) is coupling their graphql layer with their domain. There is a [popular library](https://typegraphql.com/) to do it. not saying we should move to a 1-1 mapping, but the further we get away from a tight coupling, the more the developer overhead there is. 

our v2 api already need some deprecation (some User stuff needs to be renamed to Account), so while we want to keep a API fairly stable, migration is something we have to plan anyway.

Reply:
He warns that it can be dangerous (p90):
> An annotation-based schema implementation usually uses your domain entities, objects, classes, etc. and annotates them to generate a GraphQL schema out of them. The big advantage here is that the interface layer is tiny. You don’t need hundreds of objects defining a GraphQL schema. Instead, you may simply reuse the domain entities you already have.
The danger with this approach is that it depends on your “entities” making a good API in the first place. If all you have are ActiveRecord objects closely coupled to your database schema or implementation details, an annotation-based approach risks making your resulting API not so great or stable. However, if you do have well-defined entities that represent your domain without being coupled to implementation details like the database, this might be a really great approach.


Also from the same book (p93):
> No matter how you’re building a schema, you will have to write resolvers at some point. The resolver pattern of GraphQL’s execution engine is great. It helps teams add new use cases by working on their isolated resolver and makes it simple to think about the logic of a single field. However, it can be overused. At its core, GraphQL is an API interface. It’s an interface to your domain and business logic, and in most cases, should not be the source of truth for your business.

Its an API *interface* (...) should *NOT* be source of truth!

I consider a library like [this](https://typegraphql.com/) to be poison similar to ORMs. They provide initial fast iteration at the cost of long term sustainable progress and maintainability.

If you want to revisit this we should probably bring it up @ code review to get more opinions.